lgc/interface/lgc/Pipeline.h

```diff
@@ -581,17 +582,32 @@ class Pipeline {
       CheckShaderCacheFunc;
 
   // Generate pipeline module by running patch, middle-end optimization and backend codegen passes.
-  // The output is normally ELF, but IR disassembly if an option is used to stop compilation early.
+  // The output is normally ELF, but IR assembly if an option is used to stop compilation early,
+  // or ISA assembly if -filetype=asm is specified.
   // Output is written to outStream.
-  // Like other Builder methods, on error, this calls report_fatal_error, which you can catch by setting
-  // a diagnostic handler with LLVMContext::setDiagnosticHandler.
+  //
+  // Three ways this can exit:
+  // 1. Return true for success.
+  // 2. Return false for recoverable failure because irLink asked for an "unlinked" half-pipeline and there is some
+  //    reason why the module cannot be compiled in that way. The client typically then does a whole-pipeline
+  //    compilation instead. (This cannot happen if the call to irLink did not ask for an unlinked half-pipeline.)
+  // 3. Other failures cause exit by report_fatal_error, which the client can catch by setting a diagnostic handler
+  //    with LLVMContext::setDiagnosticHandler.
```

Comment:
Other failures shouldn't really happen if the input is valid. Perhaps out-of-memory conditions, but can those really be caught with setDiagnosticHandler?

If false means "unlinked" didn't work, perhaps a Result enum would be clearer here.

Reply:
Yes, other failures tend to be because input is invalid, which llpc does already catch. I guess when they're most useful (and not caught, but just reported to stderr) is in a standalone tool.

Reply:
I don't want to do a Result enum as it seems over-the-top for the two conditions we can return here ("success" or "you need to do full pipeline compile").

LLVM generally does not provide a recoverable error mechanism from running a pass, so we can just assume that any error that wasn't supposed to happen (e.g. a pass thinks that something is invalid) will cause a report_fatal_error, or an assert, or setting your computer on fire.

Reply:
Can we just reword the comment then? "Other failures" sounds too much like the client should *expect* them and *must* handle them, when really, they're internal compiler errors that should never happen.

So I wouldn't even list them as a 3rd item here, just maybe a note that internal compiler errors are reported as assertions and/or via llvm's report_fatal_error.